packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/credential_store_state_machine.dart

```diff
@@ -153,141 +173,279 @@ class CredentialStoreStateMachine extends CredentialStoreStateMachineBase {
+        if (expirationStr != null) {
+          final secondsSinceEpoch = double.tryParse(expirationStr)?.toInt();
+          if (secondsSinceEpoch != null) {
+            expiration = DateTime.fromMillisecondsSinceEpoch(
+              secondsSinceEpoch * 1000,
+              isUtc: true,
+            );
+          }
+        }
+        awsCredentials = AWSCredentials(
+          accessKeyId,
+          secretAccessKey,
+          sessionToken,
+          expiration,
+        );
+      }
+    }
+
+    if ((userPoolTokens ?? awsCredentials ?? identityId) != null) {
+      return CredentialStoreData(
+        userPoolTokens: userPoolTokens,
+        awsCredentials: awsCredentials,
+        identityId: identityId,
+      );
+    }
+    return null;
+  }
+
+  /// Clears legacy data, it it exists.
+  Future<void> _clearLegacyCredentials(String bundleId) async {
+    final authConfig = expect<AuthConfiguration>();
+    final legacyStorageFactory = getOrCreate<LegacySecureStorageFactory>();
+    final userPoolConfig = authConfig.userPoolConfig;
+    if (userPoolConfig != null) {
+      final userPoolStorage = legacyStorageFactory.getUserPoolStorage(
+        bundleId,
+      );
+      final cognitoUserKeys = LegacyCognitoUserKeys(userPoolConfig);
+      final currentUser = await userPoolStorage.read(
+        key: cognitoUserKeys[LegacyCognitoKey.currentUser],
+      );
+      if (currentUser != null) {
+        final userPoolKeys = LegacyCognitoUserPoolKeys(
+          currentUser,
+          userPoolConfig,
+        );
+        await userPoolStorage.deleteMany([
+          userPoolKeys[LegacyCognitoUserPoolKey.accessToken],
+          userPoolKeys[LegacyCognitoUserPoolKey.refreshToken],
+          userPoolKeys[LegacyCognitoUserPoolKey.idToken],
+          cognitoUserKeys[LegacyCognitoKey.currentUser],
+        ]);
+      }
+    }
+
+    final identityPoolId = authConfig.identityPoolConfig?.poolId;
+    if (identityPoolId != null) {
+      final identityPoolStorage = legacyStorageFactory.getIdentityPoolStorage(
+        bundleId,
+        identityPoolId,
+      );
+      const identityPoolKeys = LegacyCognitoIdentityPoolKeys();
+      await identityPoolStorage.deleteMany([
+        identityPoolKeys[LegacyCognitoIdentityPoolKey.identityId],
+        identityPoolKeys[LegacyCognitoIdentityPoolKey.accessKey],
+        identityPoolKeys[LegacyCognitoIdentityPoolKey.secretKey],
+        identityPoolKeys[LegacyCognitoIdentityPoolKey.sessionKey],
+        identityPoolKeys[LegacyCognitoIdentityPoolKey.expiration],
+      ]);
+    }
+  }
+
+  @override
+  Future<void> onMigrateLegacyCredentialStore(
+    CredentialStoreMigrateLegacyCredentialStore event,
+  ) async {
+    final version = await _getVersion();
+    if (version == CredentialStoreVersion.none) {
+      final bundleIdProvider = get<BundleIdProvider>();
+      final getBundleId = bundleIdProvider?.getBundleId;
+      final bundleId = getBundleId != null ? await getBundleId() : null;
```

Comment:
I would recommend the following changes, since I don't see a good reason to return `null` from `BundleIdProvider` (and I would not have `AmplifyAuthCognitoDart` implement it or provide one to the state machine).

```dart
abstract class BundleIdProvider {
  /// Returns the bundle ID.
  FutureOr<String> getBundleId();
}
```

```suggestion
      final bundleIdProvider = get<BundleIdProvider>();
      final bundleId = await bundleIdProvider?.getBundleId();
```